src/devices/machine/mc6844.h

```diff
@@ -0,0 +1,148 @@
+// license:BSD-3-Clause
+// copyright-holders: Joakim Larsson Edström
+/***************************************************************************
+
+ Motorola 6844 emulation
+
+****************************************************************************
+                            _____   _____
+                   VSS   1 |*    \_/     | 40  E
+         *CS / Tx AKBW   2 |             | 39 *RESET
+                R / *W   3 |             | 38  DGRNT
+                    A0   4 |             | 37 *DRQ1
+                    A1   5 |             | 36 *DRQ2
+                    A2   6 |             | 35  Tx AKA
+                    A3   7 |             | 34 *TX STB
+                    A4   8 |             | 33 *IRQ / *DEND
+                    A5   9 |             | 32  Tx RQ0
+                    A6  10 |   MC6844    | 31  Tx RQ1
+                    A7  11 |             | 30  Tx RQ2
+                    A8  12 |             | 29  Tx RQ3
+                    A9  13 |             | 28  D0
+                   A10  14 |             | 27  D1
+                   A11  15 |             | 26  D2
+                   A12  16 |             | 25  D3
+                   A13  17 |             | 24  D4
+                   A14  18 |             | 23  D5
+                   A15  19 |             | 22  D6
+                   VCC  20 |_____________| 21  D7
+
+***************************************************************************/
+
+#ifndef MAME_MACHINE_MC6844_H
+#define MAME_MACHINE_MC6844_H
+
+#pragma once
+
+//**************************************************************************
+//  INTERFACE CONFIGURATION MACROS
+//**************************************************************************
+
+//**************************************************************************
+//  TYPE DEFINITIONS
+//**************************************************************************
+
+// ======================> mc6844_device
+
+class mc6844_device : public device_t, public device_execute_interface
+{
+public:
+	// construction/destruction
+	mc6844_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	auto out_int_callback() { return m_out_int_cb.bind(); }
+	auto out_txak_callback() { return m_out_txak_cb.bind(); }
+	auto out_drq1_callback() { return m_out_drq1_cb.bind(); }
+	auto out_drq2_callback() { return m_out_drq2_cb.bind(); }
+	auto in_memr_callback() { return m_in_memr_cb.bind(); }
+	auto out_memw_callback() { return m_out_memw_cb.bind(); }
+
+	// I/O operations
+	void write(offs_t offset, uint8_t data);
+	uint8_t read(offs_t offset);
+	template <unsigned CH> auto in_ior_callback() { return m_in_ior_cb[CH].bind(); }
+	template <unsigned CH> auto out_iow_callback() { return m_out_iow_cb[CH].bind(); }
+
+	template <unsigned CH> DECLARE_WRITE_LINE_MEMBER( dreq_w ) { dma_request(CH, state); }
+
+	DECLARE_WRITE_LINE_MEMBER( dgrnt_w ){ m_dgrnt = state; trigger(1); }
```

Comment:
You’re intermixing configuration and live methods.

Also, what are you doing to protect against time travel effects when inputs are updated from a different scheduled device's time domain?

Reply:
I understand what you say but not what the best way to fix it would be. Something like input merger or what?

Reply:
You missed the point completely  You've got the configuration methods for the callbacks, followed by `write` and `read` which are live methods, followed by more configuration methods for callbacks.  Grouping the configuration methods together makes it easier to understand the code.

Reply:
ahh ok, I reorder them

I was in addition already concerned about the trigger call in the template but that might be ok then from a time domain perspective